### ISSUE TITLE
Thread-ConsumeBidirectionalStream exiting  and not re establishing stream -  collection watch

### ISSUE THREAD

**Issue by BarisicLuka** [googleapis/python-firestore#399]
#### Environment details

  - OS type and version: Debian GNU/Linux 10 (buster)
  - Python version: `Python 3.7.3`
  - pip version: `pip 21.0.1`
  - `google-cloud-firestore` version: `Version: 2.1.3`

#### Context
I created a collection watch that listens to a firestore collection in a firebase project and redirects the data in another format in a firestore collection of another firebase project.
It worked fine for a while then started to stop working when I launched two instances  of the same script (to  redirect the data to my dev and prod environment). When I run the scrip there are no problem for a while and at some point I get the log :
``INFO  | google.api_core.bidi| _thread_main()| L676 | Thread-ConsumeBidirectionalStream exiting| call_trace=/path/venv/lib/python3.7/site-packages/google/api_core/bidi.py| L676 ``
which, from my understanding, should be followed by a reinitialization of the thread but  the only things that I see are some unsuccessful attempts as nothing happens after those: 
`` INFO  | google.api_core.bidi| _reopen()| L487 | Re-established stream| call_trace=/path/venv/lib/python3.7/site-packages/google/api_core/bidi.py| L487 ``

When this happens,  the  python process is not killed, it is still running and periodically tries to re establish the stream to firestore but does not succeed.

I looked for some similar errors on github but none of the mentioned methods fix this behavior, not even updating the google-cloud-firestore version. Also here is the [stackverflow question](https://stackoverflow.com/questions/67989237/firestore-collection-snapshot-exiting-and-never-starting-again-thread-consume) that I created.

The problem  used to  happen once every so often  so I managed  to patch it by reading the logs every minutes to see it the process stoped but yesterday it kept happening for a long period. After I ran the script, whether I executed it on my dev end, prod env or both it  stopped after 1 to 3 minutes.

#### Steps to reproduce

It's pretty difficult to tell you how to reproduce it as I have no clue what may cause this behavior.

#### Code example

Here is my snapshot (dynamic.py)
```python
import datetime
import threading

from google.api_core.exceptions import NotFound

from databases import (ORIGIN_DB, DESTINATION_DB)
from config import (OBJECT_COLLECTION)
# Create an Event for notifying main thread.
callback_done = threading.Event()


# Create a callback on_snapshot function to capture changes
def redirect_data(col_snapshot, changes, read_time):
    for change in changes:
        if change.type.name == 'ADDED':
            ts = datetime.datetime.now().astimezone()
            document_dict = change.document.to_dict()
            document_dict['updated_at'] = ts
            for object_id, object_data in document_dict.items():
                object_data['created_at'] = ts
                DESTINATION_DB.collection(OBJECT_COLLECTION) \
                    .document(object_id) \
                    .set(object_data)
        # Document deletion
        if change.type.name == 'REMOVED':
            document_dict = change.document.to_dict()
            for object_id in document_dict.keys():
                DESTINATION_DB.collection(OBJECT_COLLECTION) \
                    .document(object_id) \
                    .delete()
        # Document update
        if change.type.name == 'MODIFIED':
            ts = datetime.datetime.now().astimezone()
            document_dict = change.document.to_dict()
            document_dict['updated_at'] = ts
            for object_id, object_data in document_dict.items():
                # If the document does not exist we create it
                try :
                    DESTINATION_DB.collection(OBJECT_COLLECTION) \
                        .document(object_id) \
                        .update(object_data)
                except NotFound as e:
                    DESTINATION_DB.collection(OBJECT_COLLECTION) \
                        .document(object_id) \
                        .set(object_data)
    callback_done.set()

col_query = ORIGIN_DB.collection('my_collection')

# Watch the collection query
query_watch = col_query.on_snapshot(redirect_data)
```

and here is my main.py

```python3
from config import LOGGERS_CONFIG

logging.config.dictConfig(LOGGERS_CONFIG)
logger = logging.getLogger()


if __name__ == '__main__':
    from dynamic import query_watch 
    while True:
        pass

```
#### Stack trace
Here are the logs that I had yesterday 
```
2021-07-15 17:13:22,466| INFO  | root| redirect_data()| L89  | Updating slot : OB1yqwmLrAgTPcWXtP3J
2021-07-15 17:13:22,568| INFO  | root| redirect_data()| L89  | Updating slot : 8fhJmHyZUaJmQ4Ul7n6c
2021-07-15 17:13:22,740| INFO  | root| redirect_data()| L89  | Updating slot : svTOjjciiZE9PgrR4WrE
2021-07-15 17:13:22,844| INFO  | root| redirect_data()| L89  | Updating slot : h6AxRrf6chhLzruOLxjP
2021-07-15 17:13:23,542| INFO  | google.api_core.bidi| _thread_main()| L676 | Thread-ConsumeBidirectionalStream exiting| call_trace=/path/venv/lib/python3.7/site-packages/google/api_core/bidi.py| L676 
```

And after restarting it I get : 

```
2021-07-15 17:15:57,102| INFO  | root| redirect_data()| L89  | Updating slot : Ccy2VeFExcieqQhY3nMb
2021-07-15 17:15:57,232| INFO  | root| redirect_data()| L89  | Updating slot : xMPTqs5Lo0zIsGkcgGo1
2021-07-15 17:15:57,348| INFO  | root| redirect_data()| L89  | Updating slot : scEcZfi7RsVRbT9C8O8p
2021-07-15 17:15:57,522| INFO  | root| redirect_data()| L89  | Updating slot : 7MH6rbKFHRuTWI1gWI2z
2021-07-15 17:15:57,958| INFO  | google.api_core.bidi| _thread_main()| L676 | Thread-ConsumeBidirectionalStream exiting| call_trace=/path/venv/lib/python3.7/site-packages/google/api_core/bidi.py| L676 
```

Yesterday was the worst period, it happened every 1-3 minutes from 4:20 PM to 6:30PM CEST approximatively.

I would be glad to know if there are any means to fix or if what I am trying to do goes against one of Firestore concept. Also I know that using google cloud functions could eventually  fix it but it is more convenient for me to use python.

Another note is that it isn't the only query watch for which I had this problem, it happened twice for a query watch that  was simply listening to a collection to take action upon the received data WITHIN the same firebase  project.

Thank you in advance

**Comment by craiglabenz** [comment]
Thanks for reporting this issue, @BarisicLuka! Sadly, issues with `watch` are famously hard to reproduce, but you seeing it every 3 minutes is very interesting. I've spent some time attempting to recreate this issue, but so far am getting bogged down in attempting even a primitive recreate with the script you provided - mostly because it seems you've written it to understand your data structure, which I don't know :)

That said, if you could provide some extra info, we can try to recreate the full scenario. It would help to see the contents of your `databases.py` and `config.py` files, as well as what your documents in question look like.

**Comment by craiglabenz** [comment]
#175 and #18 are potentially similar issues.

**Comment by BarisicLuka** [comment]
Hey @craiglabenz,

Thanks for looking into my issue ! 
You're right, the code I provided is an adapted version of the one I used as I changed a couple of things so that I can share it openly. I shared it just for you to get a glimpse of what I was doing but haven't tested it, so shame on me. 
I will work on some useful and running example for you to be able to try and recreate it. Hopefully I'll be able to provide it to you next week. 

It is very similar to #18 indeed, especially because I also run two watch on the same collection. Unfortunately upgrading my version did not change anything.

**Comment by tseaver** [comment]
@BarisicLuka can you perhaps run your script with `logging.basicConfig(level=logging.DEBUG)` enabled?  We should be able to see the reason for the RPC terminating with that info.

**Comment by BarisicLuka** [comment]
Hello @tseaver 

I got caught up in my work and forgot to provide a better exemple, sorry about that.
We changed the way we were doing things and do not use those functions anymore and using them for some tests would impact our production environment so I cannot really run my scripts as of now.

The problem appeared when I was copying data from one Firestore project to two others but did not appear when I was copying data  from one Firestore  project to a single other.

I know this is not helpful but there's little I can do to recreate it as of now.
